es7/2015-11/nov-18.md

```diff
@@ -0,0 +1,1016 @@
+
+
+### Improving consistency of @@species (Kevin Smith)
+
+KS: on map and set, they are not used in the spec themselves, But subclasses should be able to override
+
+KS: usage patterns, instance methods, that want to return related subclass.
+
+KS: Promise.all and Promise.race are the only ones that use this
+
+YK: race/all are combinators Array.from is a custom constructors:
+
+KS: it would be good to define what the usage patterns are
+
+AWB: ..
+
+DD: I disagree, resolve/all/race is more “casting” and species should not be used.
+
+AWB: I agree the return value should be using species, as the user is specifying.
+
+DD: the argument values can be anything, and they are to be cast to the RacePromise
+
+AWB: but why
+
+DD: an example …Could be… CancellablePromise.race(arrayOfMiscPromises) likely wants cast its input.
+
+YK: those examples are dubious
+
+SP: lets us InstrumentedPromise as an example
+
+DD: ok lets use that one, InstrumentedPromise.all(mixOfPromises) casts all inputs to InstrumentPromises
+
+DD: auto-casting is the intent.
+
+AWB: that works great for the case, where you want to cast
+
+AWB: but it breaks the case where you don’t.
+
+DD: I believe it was a mistake, likely my own mis-understanding. I don’t believe it makes sense.
+
+DE: zepto took a literal array that and swaped out its constructor. If that was new’d it wouldn’t work. `Array['@@species']` create, has two checks. this.constructor if thats not an object, it goes back to the default `ArrayCreate['@@species']` of fallback. The critical original issues (based on the notes) was WebCompat
+
+YK: @@species has another precednt
+
+DD: A solution was needed, this pattern appealed to others. I believe it was misused, and misunderstood.
+
+DD: it expressed a use-case for arrays, that are hypothetically use-full.
+
+AWB: avoid nodelist instead get an array out of it.
+
+KS: I would prefer, i would not want to argue removing it. I would like understand how it ought to be used.
+
+YK: why is it on map/set
+
+AWB: someone argued it should be for the future, for example adding filter/map
+
+AWB: if we wanted to provide filter to math, a subclass to math. This would allow them to change species
+
+YK: we should add it on-demand.
+
+DD: we should move away from @@species on the promise constructor
+
+DD: we should discover a future strategy. `Function.prototype['@@species'] = this`
+
+AWB: not sure why we didn’t think about this sooner
+
+DH: whats the best next step for a change like that
+
+YK/DD: PR
+
+DH: should I file a bug if i don’t have time
+
+DD/KS: bug
+
+YK: should we remove it from map
+
+DD: lets remove it from everything, and put it on `Function.prototype['@@species'] = this;`
+
+AWB: it was added for array instance extensibility.
+
+DD: that would break zepto
+
+DE: Object.__proto__ is Function.prototype, which have a species returning this, which breaks the web.
```

Comment:
![2015-12-23_15-37-02](https://cloud.githubusercontent.com/assets/19714/11972139/098c4f5e-a98b-11e5-9f29-19b16c85bfc1.jpg)

should be 

    `Object.__proto__`